### PR TITLE
[FIRRTL] AdvancedLayerSink: don't sink instances of mods with port annos

### PR DESCRIPTION
**Pull request by rwy7**
If a module-like has port annotations, mark it as an effectful module, which will prevent the pass from moving or deleting its instances later on. We can't sink instance when we don't understand what port annotations mean.

### CODE REVIEW

File: test/Dialect/FIRRTL/advanced-layer-sink.mlir [hunk 1]
```diff
@@ -715,3 +715,26 @@ firrtl.circuit "Sub" {
...
+    attributes {
+      portAnnotations = [
+        [{class = "circt.FullResetAnnotation", resetType = "async"}]
+      ]
+    }
+  {}
+
+  // CHECK: firrtl.module @DoNotSinkInstanceOfModuleWithPortAnno
+  firrtl.module @DoNotSinkInstanceOfModuleWithPortAnno() {
+    // CHECK-NEXT: firrtl.instance foo @ModuleWithPortAnn
```
Comment (dtzSiFive) [inline test/Dialect/FIRRTL/advanced-layer-sink.mlir:733]:
```suggestion
    // CHECK-NEXT: firrtl.instance foo @ModuleWithPortAnno
```

**Review by dtzSiFive — APPROVED** [review]
LGTM!

File: lib/Dialect/FIRRTL/Transforms/AdvancedLayerSink.cpp [hunk 2]
```diff
@@ -111,9 +111,18 @@ class EffectInfo {
...
   void update(FModuleLike moduleOp) {
+    // If the module op has any annotations, then we pretend the module contains
+    // some kind of important effect, so that we cannot sink its instances.
     if (!AnnotationSet(moduleOp).empty())
       return markEffectful(moduleOp);
+
+    for (auto annos : moduleOp.getPortAnnotations()) {
+      if (!AnnotationSet(cast<ArrayAttr>(annos)).empty())
+        return markEffectful(moduleOp);
+    }
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/AdvancedLayerSink.cpp:124]: Mega-nit: 
```suggestion
    for (auto annos : moduleOp.getPortAnnotations())
      if (!AnnotationSet(cast<ArrayAttr>(annos)).empty())
        return markEffectful(moduleOp);
```

Suggestion (the `AnnotationSet` is going to be thrown away): 

```suggestion
    for (auto annos : moduleOp.getPortAnnotations()) {
      if (!cast<ArrayAttr>(annos).empty())
        return markEffectful(moduleOp);
    }
```

**Review by seldridge — APPROVED** [review]
LGTM

**Pull request merged**